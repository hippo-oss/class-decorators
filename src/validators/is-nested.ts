import { composePropertyDecorators, IsNestedOptions } from '@hippo-oss/dto-decorators';
import {
    ValidateNested,
    ValidatorConstraintInterface,
    Validate,
    ValidationArguments,
    ValidatorConstraint,
} from 'class-validator';

import { createBasePropertyDecorators } from './base';

@ValidatorConstraint({ name: 'IsOneOf', async: false })
export class IsOneOf implements ValidatorConstraintInterface {
    validate(value: unknown, args: ValidationArguments) {
        // allow to be handled by normal ValidateNested
        if (typeof value !== 'object' || value === null) {
            return true;
        }

        const definedProperties = Object.values(value)
            .filter((property) => property !== undefined && property !== null);

        return definedProperties.length === args.constraints[0];
    }

    defaultMessage(args: ValidationArguments) {
        // args.constraints[0] should always exist and be a number
        const constraint = args.constraints[0] as number;

        // args.value should always be an object here based on the validate function
        const value = args.value as Record<string, unknown>;
        const definedProperties = Object.keys(value).filter(
            (key) => value[key] !== undefined && value[key] !== null,
        );

        return `Exactly ${constraint} of the properties of ${args.property} must be set,`
        + ` you are currently setting properties: [${definedProperties.join(', ')}]`;
    }
}

export function IsNested(options: IsNestedOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate nested
        ValidateNested({ each: !!options.isArray }),

        // maybe: add if only X properties can be set
        options?.hasXPropertiesDefined !== undefined ? Validate(IsOneOf, [options.hasXPropertiesDefined]) : undefined,
    ]);
}
