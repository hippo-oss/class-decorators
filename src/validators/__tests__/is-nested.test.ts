import { validateSync } from 'class-validator';

import { IsNested } from '../is-nested';

describe('IsNested', () => {
    class Child {
    }

    it('allows optional nested to be omitted', () => {
        class Fixture {
            @IsNested({
                optional: true,
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable nested to be omitted', () => {
        class Fixture {
            @IsNested({
                nullable: true,
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required nested to be omitted', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-nested', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        obj.property = 'foo' as unknown as Child;
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows nested', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        obj.property = new Child();

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    describe('IsOneOf', () => {
        it('does not allow more parameters than expected', () => {
            class Fixture {
                @IsNested({
                    type: Child,
                    hasXPropertiesDefined: 1,
                })
                property!: Child;
            }

            const obj = new Fixture();
            obj.property = { test1: 'test', test2: 'test2' };

            const errors = validateSync(obj);
            expect(errors).toHaveLength(1);
            expect(errors).toMatchSnapshot();
        });
        it('does not allow less parameters than expected', () => {
            class Fixture {
                @IsNested({
                    type: Child,
                    hasXPropertiesDefined: 1,
                })
                property!: Child;
            }

            const obj = new Fixture();
            obj.property = { test1: undefined, test2: null };

            const errors = validateSync(obj);
            expect(errors).toHaveLength(1);
            expect(errors).toMatchSnapshot();
        });
    });
});
