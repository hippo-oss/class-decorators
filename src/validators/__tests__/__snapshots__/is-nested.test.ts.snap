// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsNested IsOneOf does not allow less parameters than expected 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "IsOneOf": "Exactly 1 of the properties of property must be set, you are currently setting properties: []",
    },
    "property": "property",
    "target": Fixture {
      "property": Object {
        "test1": undefined,
        "test2": null,
      },
    },
    "value": Object {
      "test1": undefined,
      "test2": null,
    },
  },
]
`;

exports[`IsNested does not allow non-nested 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "constraints": Object {
          "nestedValidation": "nested property property must be either object or array",
        },
        "property": "property",
        "target": [Function],
        "value": "foo",
      },
    ],
    "property": "property",
    "target": Fixture {
      "property": "foo",
    },
    "value": "foo",
  },
]
`;

exports[`IsNested does not allow required nested to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "property should not be null or undefined",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
